ggml-rpc.cpp

```diff
@@ -96,27 +96,37 @@ static ggml_guid_t ggml_backend_rpc_guid() {
     return &guid;
 }
 
-struct ggml_backend_rpc_buffer_type_context {
+struct rpc_backend {
+    int ref_count;
```

Comment:
maybe we can try something link intrusive shared pointer: [std::enable_shared_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

Reply:
Yes, I think that the correct way to do this would be to change the pointers in `instances` to `weak_ptr`, then the backend will be automatically freed by `shared_ptr` on the last instance. `create_rpc_backend` would need to check if the `weak_ptr` in instances is still alive.

Reply:
yeah, that could be the best, inside the `instances` we maintain a `weak_ptr<rpc_backend>`, without increase the reference count, and then at the `create_rpc_backend` we obtain `shared_ptr` from `weak_ptr`.
this also save the `erase` call at `free_rpc_backend`

Reply:
Unfortunately it is not that simple. We keep a backend reference in `ggml_backend_rpc_buffer_type_context` and we use this reference when allocating new buffers. If we free this reference in `ggml_backend_rpc_free()` then we won't be able to allocate new buffers.

Reply:
IIRC, we have several structure have a reference to `rpc_backend` here:
1. `ggml_backend_rpc_buffer_type_context`
2. `ggml_backend_rpc_buffer_context`
3. `ggml_backend_rpc_context` 
4. `instances`.

first 1-3 should have a strong reference to let them allocate buffer after `ggml_backend_rpc_free` call explicitly - that's the case you mention.
and for the 4, i think we can be make it as `weak_ptr` since we'll have `ggml_backend_rpc_context` which life-span is expected longer than the `rpc_backend`?

Reply:
When do we free the reference which is kept in `ggml_backend_rpc_buffer_type_context`? If you free this reference, then you can no longer allocate buffers. If you don't free this reference, then the backend will never be freed because of this reference.

I don't see how to solve this without explicit reference count. I also don't see how `weak_ptr` can help here.

Reply:
Buffer types are never freed, I think it would complicate the user code too much to require buffer types to be freed, for no real benefit. In the RPC backend, buffer types only need a connection during initialization and during buffer allocation. You can make a connection during initialization to obtain the alignment and max size, drop it, and only open it again once a buffer is allocated, and then rely on the `shared_ptr` to keep it alive until all buffers and backend instances have been freed.

This is probably complicated because the RPC backend uses a `ggml_backend` instance to represent a connection, but connections and `ggml_backend` backend instances should probably be different concepts.

Reply:
ok, let's say that the connection is not tied to the backend instance and we have a separate entity (e.g. `rpc_connection`) which is being referenced with `shared_ptr`. How do we obtain a `shared_ptr` to the connection when a new buffer is allocated?

Reply:
I think something like this should work:

```C++
std::shared_ptr<connection_t> get_connection(const std::string & endpoint) {
    static std::map<std::string, std::weak_ptr<connection_t>> connections;

    auto it = connections.find(endpoint);
    if (it != connections.end()) {
        if (auto connection = it->second.lock()) {
            return connection;
        }
    }

    auto connection = std::make_shared<connection_t>(endpoint);
    connections[endpoint] = connection;
    return connection;
}
```

Use this function when you need to create a new buffer or backend instance. You can also use it during initialization of the buffer type, just use a local `shared_ptr` in the function only, but don't keep the instance.